New depart

RemyVase opened the pull request:
Yo regarder les commits, j'ai changé des trucs et limite repartis de 0 mais ca fait plus propre. Mémoire partagée compris reste à l'implanté

RemyVase committed:
Test printf du utils

En gros j'ai refais un utils dans lequel je met beaucoup de foncvtion pour allégé le main. Ici j'ai fais juste des printf je vais tenter de mofid par après. Je dois refaire des forks etc.

RemyVase committed:
Create utils

Maintenant timeFormat() affiche le temps en xxminxxsecxxx, Ajout de fonction pour généré les temps des secteurs de manière plus propre dans le main, j'ai déjà mis une fonction de tri pour plus grand plus petit pour bestSect. Truc pomper du net. Et ajout de fonction pour pit et out ca sera aussi plus propre :)

ATHOOS committed:
memoire partagee wip

memoire pas encore terminee

ATHOOS committed:
memoire partagee update

memoire partagee cree et associee entre processus
encore quelque soucis en terme de suppression de la memoire (si on ne supprime pas la memoire, le temps total ne se reinitialise pas entre differents run du main)

ATHOOS commented:
J'ai avancé dans la memoire partagee, j'ai reussi a ecrire au travers et a lire. (J'ai un peu suivie le mode de fonctionnement du projet de Brieux et Maxence pour ça). J'ai aussi supprimé les fonctions inutiles du utils.h (style genS1, genS2, genere_chance, etc). Le code est un peu plus clair je pense, il faudra encore que je commente ce qu'il faut afin de rendre ca encore plus clair.

amostin commented:
GG :smile:

ATHOOS committed:
memoire partagee finalisee

correction de la reinitialisation de la memoire, et derniere suppression des fonctions "inutilisees" du utils.h

ATHOOS committed:
update display

ajout de deux fonctions d'affichage, permettra d'afficher la grille de départ et le podium de fin

ATHOOS commented:
Alors j'ai juste rajouté deux petites fonctions qui permettent d'afficher la grille de départ, et le podium de fin. (Ouais je sais ça sert à rien, mais j'avais envie de faire un truc pas trop dur ce soir xD)

ATHOOS committed:
edit shmget

petite correction

ATHOOS commented:
![Capture d’écran de 2019-07-28 01-54-23](https://user-images.githubusercontent.com/43479148/62000648-62e55080-b0dc-11e9-9955-52b4c357f1ae.png)
![Capture d’écran de 2019-07-28 01-54-30](https://user-images.githubusercontent.com/43479148/62000649-6547aa80-b0dc-11e9-9b78-ec98bbf6bb0d.png)

ATHOOS committed:
changement printf

changement de grid en depart

ATHOOS committed:
test des semaphores

semaphore fonctionnels, reste a terminer l'adaptation au projet

ATHOOS commented:
Bon pour ce qui est des sémaphores.
Il a fallu créer une nouvelle structure dédiée spécialement aux sémaphores. Elle s'appelle "Buffer", elle contient plusieurs champs spécifiques à la manière qu'on a utilisé pour l'écritture et la lecture. 
On s'est rendu compte qu'il serait plus simple de travailler avec une "queue" plutôt qu'avec une "stack".
Une "queue" suit le principe de First In, First Out, tandis que la stack suit le principe de First In, Last Out. 
![Queue-program-enqueue-dequeue](https://user-images.githubusercontent.com/43479148/62156466-5d218200-b30b-11e9-8f2d-4477929fa601.jpg)
Cette image illustre bien le fonctionnement de la queue. Juste une particularité, on a fait une sorte de boucle, afin de ne pas atteindre un moment où aucune place n'est disponible.
Pour le fonctionnement des sémaphores en eux-mêmes :
- on crée la structure dédiée aux sémaphores 
 `typedef struct Buffer{
    int shmid_buffer;
    int size;
    int maxsize;
    int last;
    int first;
    sem_t mutex;
    sem_t full;
    sem_t empty;
}buffer;`
- on initialise la structure des sémaphores grâce à la fonction init_buff(). On initialise les balises de début et de fin ainsi que la taille pour la queue. On initialise également les différents sémaphores "mutex", "empty", "full".
- pour l'écriture dans le mémoire partagée, le processus se passe en plusieurs phase (int insert())
         1. on attend qu'une place se libère `sem_wait(&b->empty);`
         2. on vérifie que la ressource n'est pas bloquée par un mutex `sem_wait(&b->mutex);` et on la bloque afin qu'aucun autre processus n'y ait accès
         3. on écrit les données précédement générée dans cet espace mémoire
         4. on débloque le sémaphore mutex `sem_post(&b->mutex);`
         5. on "redonne" la place prise au sémaphore `sem_post(&b->full);`
- en ce qui concerne la lecture (voiture rem()), on exécute le déroulement opposé, on vérifie qu'un emplacement est rempli, on bloque, on lit, on débloque, et on rend la place vide

ATHOOS committed:
update sem and shm

suppression de la deuxieme memoire, et update des semaphore

ATHOOS commented:
Nouvelle version du code dispo dans le fichier testSem.c
Il y a aussi un truc important a noter, c'est qu'on ne compile plus de la même manière maintenant. Il faut ajouter l'option lpthread à la compilation. Ce qui donne par exemple:
`gcc -o test testSem.c -lpthread`

ATHOOS committed:
nouvelle update sem

Tjr fix un probleme sur le taille du buffer des semaphore, on aurait beoins d'une plus grande taille mais impossible de faire ca, on se retrouve avec une erreur de segmentation

ATHOOS committed:
semaphore final

semaphore corrigés et corrects

ATHOOS commented:
Voila juste le resultat maintenant : 
![Capture d’écran de 2019-08-01 02-17-19](https://user-images.githubusercontent.com/43479148/62256710-a6a0c880-b402-11e9-93a3-9a8e97e4176a.png)
20 voitures qui tournent sur 100 tours sans problème

ATHOOS commented:
Et tout est affiché dans l'ordre des tours, ils nous reste à juste à refresh correctement notre affichage.
